### ISSUE TITLE
Improve multiple dispatch

### ISSUE THREAD

**Issue by evhub** [evhub/coconut#92]
Currently, multiple dispatch using `def` and `case` is rather clunky. The suggestion is to replace
```
def some_func(x):
    case x:
        match _ is int:
            do_stuff_for_int(x)
        match _ is str:
            do_stuff_for_str(x)
```
with
```
case def some_func:
    match x is int:
        do_stuff_for_int(x)
    match x is str:
        do_stuff_for_str(x)
```

**Comment by datnamer** [comment]
How about a way to register new specializations of a function? Like this library: https://github.com/mrocklin/multipledispatch 

That way, we can write modular, generic methods code that can be extended by others, instead of writing one monolithic function. 

Preferably the API would use type hints: https://github.com/un-def/polypie

**Comment by evhub** [comment]
Interesting idea—I'll definitely look into that and see what I can do.

**Comment by datnamer** [comment]
The inspiration is from Julia. 

This is a paper with information on the design of multi dispatch and how it plays into abstraction, specialization, performance etc : https://arxiv.org/abs/1411.1607 ...if you feel have time for extra yet fairly interesting reading.

Its useful for general computing because a core library can define a type or an interface and other libraries can add new methods without subclassing.

**Comment by evhub** [comment]
How about this:
```
def func(x is int) = x + 1
@append_case(func)
def func(x is str) = x + "!"
func(5) == 6
func("5") == "5!"
```

**Comment by evhub** [comment]
Currently live in develop.

**Comment by evhub** [comment]
@datnamer Released in v1.1.0!

**Comment by datnamer** [comment]
Awesome!
